lib/pos_snapshot_entries.go

```diff
@@ -569,28 +565,60 @@ func DBPutSnapshotValidatorSetEntryWithTxn(
 ) error {
 	if validatorEntry == nil {
 		// This should never happen but is a sanity check.
-		glog.Errorf("DBPutSnapshotValidatorSetEntryWithTxn: called with nil ValidatorEntry, this should never happen")
+		glog.Errorf("DBUpdateSnapshotValidatorSetEntryWithTxn: called with nil ValidatorEntry, this should never happen")
 		return nil
 	}
 
-	// Put the ValidatorEntry in the SnapshotSetByPKID index.
-	key := DBKeyForSnapshotValidatorSetByPKID(validatorEntry, snapshotAtEpochNumber)
-	if err := DBSetWithTxn(txn, snap, key, EncodeToBytes(blockHeight, validatorEntry), eventManager); err != nil {
+	// Look up the existing entry
+	dbEntry, err := DBGetSnapshotValidatorSetEntryByPKIDWithTxn(
+		txn, snap, validatorEntry.ValidatorPKID, snapshotAtEpochNumber)
+	if err != nil {
 		return errors.Wrapf(
-			err,
-			"DBPutSnapshotValidatorSetEntryWithTxn: problem putting ValidatorEntry in the SnapshotValidatorByPKID index: ",
+			err, "DBUpdateSnapshotValidatorSetEntryWithTxn: problem retrieving ValidatorEntry for PKID %v: ",
+			validatorEntry.ValidatorPKID,
 		)
 	}
+	dbEntryBytes := EncodeToBytes(blockHeight, dbEntry)
 
-	// Put the ValidatorPKID in the SnapshotValidatorByStatusAndStakeAmount index.
-	key = DBKeyForSnapshotValidatorSetByStakeAmount(validatorEntry, snapshotAtEpochNumber)
-	if err := DBSetWithTxn(txn, snap, key, EncodeToBytes(blockHeight, validatorEntry.ValidatorPKID), eventManager); err != nil {
+	entryToWriteBytes := EncodeToBytes(blockHeight, validatorEntry)
+	// If the entry in the db is the same as the entry we want to write, then no need to do anything.
+	if bytes.Equal(dbEntryBytes, entryToWriteBytes) {
+		return nil
+	}
+
+	// Put the ValidatorEntry in the SnapshotSetByPKID index.
+	key := DBKeyForSnapshotValidatorSetByPKID(validatorEntry, snapshotAtEpochNumber)
+	if err := DBSetWithTxn(txn, snap, key, EncodeToBytes(blockHeight, validatorEntry), eventManager); err != nil {
 		return errors.Wrapf(
 			err,
-			"DBPutSnapshotValidatorSetEntryWithTxn: problem putting ValidatorPKID in the SnapshotValidatorByStake index: ",
+			"DBUpdateSnapshotValidatorSetEntryWithTxn: problem putting ValidatorEntry in the SnapshotValidatorByPKID index: ",
 		)
 	}
 
+	// Only update the PrefixSnapshotValidatorSetByStakeAmount index if the stake amount has changed.
+	if dbEntry == nil || dbEntry.TotalStakeAmountNanos.Cmp(validatorEntry.TotalStakeAmountNanos) != 0 {
```

Comment:
This one is fine. The validatorEntry will never change once snapshotted.

Reply:
I think it's okay to get rid of the changes here since once we snapshot, the values will never change. we only flush snapshot keys once.